### DISCUSSION TITLE
Experimental logs route

### DISCUSSION THREAD

**Discussion by irevoire** [meilisearch/product#721] [Experimental features]
<table>
    <thead>
        <tr>
            <th>Feature name</th>
  			<th>Enable when launching</th>
  			<th>Enable at runtime</th>
  			<th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Get logs through a route</td>
           <td>

```
./meilisearch --experimental-enable-logs-route
```
  
</td>            
<td>

```
curl \
  -X PATCH 'http://localhost:7700/experimental-features/' \
  -H 'Content-Type: application/json'  \
--data-binary '{
    "logsRoute": true
  }'
```
  
 </td>
  			<td>Create a new route under `/logs/stream` that streams logs to the client</td>
        </tr>
    </tbody>
</table>

Meilisearch v1.7.0 (released on March 11, 2024) introduces a new route to get logs.

## Experimental feature abstract

Enabling the experimental feature will make a new `/logs/stream` route available in Meilisearch that lets you configure which logs you want and their format.
Analyzing those logs can give great information on what's happening inside the engine and how the Meilisearch team can improve the indexing performances.
It also makes a new `/logs/stderr` route available in Meilisearch that lets you configure the logs you want on your console at runtime.

### How to use the feature?

1. Enable the feature through the CLI parameter `--experimental-enable-logs-route` or call the `PATCH /experimental-features` passing `"logsRoute": true`.
2. Call the `/logs/stream` route with your parameters:
```
curl \
	-X POST http://localhost:7700/logs/stream \
	-H 'Content-Type: application/json' \
	--data-binary '{
	    "mode": "human",
            "target": "index_scheduler=trace"
    }'
```
3. When you stop listening on the route, Meilisearch should drop the connection automatically at some point, but that may take a while. If you want to stop the connection straight away, you can call the `/logs/stream` route again but with the `DELETE` verb:
```
curl \
        -X DELETE http://localhost:7700/logs/stream
```

----

Or you can simply update the log level of the logs outputted to stderr by sending:
```
curl \
	-X POST http://localhost:7700/logs/stderr \
	-H 'Content-Type: application/json' \
	--data-binary '{
            "target": "index_scheduler=trace"
    }'
```

-------

The available `mode`s are:
- `human`: Those are the same logs you see in the console.
- `profile`: This mode profiles the time spent in each instrumented function of meilisearch. It can help to understand performance issues

The `target` parameter lets you specify which parts of the code you want to inspect and is formatted like that: `code_part=log_level,code_part=log_level`. See the following valid examples:
- `milli=trace,index_scheduler=info,actix_web=off`
- `milli` is a valid shortcut to enable the `info` logs for milli
- `trace` is another shortcut to enable the trace level for everything
- If you don’t specify anything, `info` is used as the default value
- If you specify an empty string, a reminder of the format is returned in an error message

💡 More information [here](https://www.notion.so/Logs-trace-changes-in-v1-7-0-6dd111ecd66448aba30d2f3cef212749?d=b1dc11dd2e564ddc949938efe1eb7502&pvs=4#12d8673a19df4cfda80d9dc331df81de)

## Why is this feature not stable yet?

The main purpose of this route right now is to help people debug issues in production without restarting their instance and profile Meilisearch.
We expect the route to evolve over time as we discover issues and information we're missing.

🗣️  **You are welcome to give feedback or ask any question on its usage; we are eager to collect feedback on the feature**

-------

A few notes on the feature:
- You can’t have two clients listening at the same time on the route of the logs
- To access the route, you need an API-key with at least the `metrics.get` action
- I noticed that streaming logs didn’t work with `httpie`. The client should support HTTP streams.

**Comment by sanders41** [comment]
I was experimenting with this and I think I had the same issue `httpie` has. It _kind of_ works. With `httpie` if you run it with the `--stream` flag it looks frozen, but it is working. If you kill the stream with `curl -X DELETE http://localhost:7700/logs/stream` while `httpie` is running the logs will print after the stream stops. It seems to wait for the stream to end to print it all at once.

I'm not sure if any of this helps and I don't know anything you can do about it, but thought I'd pass on what I have found :smile:

**Reply by sanders41** [reply]
I doubt this is helpful, and it for sure won't help with httpie, but in the interest of providing all the information I have this is the best work around I have come up with so far. It only prints the results after the sleep, but at least you get the results.

```py
import asyncio

from httpx import AsyncClient, StreamError


async def sample_logs() -> None:
    async def process_stream(client: AsyncClient) -> None:
        payload = {"mode": "human", "target": "index_scheduler=info"}
        async with client.stream(
            "post",
            "logs/stream",
            json=payload,
            headers={"Content-Type": "application/json"},
        ) as response:
            if response.status_code != 200:
                a = await response.aread()
                raise StreamError(a.decode())
            async for line in response.aiter_lines():
                print(line)

    async with AsyncClient(
        base_url="http://127.0.0.1:7700",
        headers={"Authorization": "Bearer masterKey"},
        timeout=None,
    ) as client:
        task = asyncio.create_task(process_stream(client))
        await asyncio.sleep(10)
        await client.delete("logs/stream")
        await task


async def main() -> int:
    await sample_logs()

    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
```

**Reply by irevoire** [reply]
Oh yes, nice to know, thanks!

**Comment by macraig** [comment]
**Update:** [v1.7](https://github.com/meilisearch/meilisearch/releases/tag/v1.7.0) has been released 🐇

**Comment by agardnerIT** [comment]
Can the logs be dumped to a file instead of streamed? I want to pull the logs into an external storage via the OpenTelemetry filelog receiver

**Reply by irevoire** [reply]
Hey, I don't think you need the logs route to do that.
Piping stderr to a file when running the engine should do it!
```
./meilisearch --your-options-here 2> logs_file
```

**Reply by agardnerIT** [reply]
I tried this and I can see the log file, but it's empty. I'm running Podman on windows.

```
version: '3.8'

volumes:
  meili_storage:

services:
  meilisearch:
    env_file: .env
    # logging:
    #   driver: "json-file"
    image: getmeili/meilisearch:v1.15
    restart: always
    environment:
      - MEILI_ENV=development
      - MEILI_LOG_LEVEL=TRACE
      - MEILI_NO_ANALYTICS=true
      - MEILI_EXPERIMENTAL_LOGS_MODE=JSON
      - MEILI_EXPERIMENTAL_ENABLE_METRICS=true
    ports:
      - 7700:7700
    volumes:
      - meili_storage:/meili_data
    command: ["/bin/sh", "-c", "meilisearch > /meili_data/output.log"]
 ```
 
 Running under WSL, when I set the `logging.driver` to `json-file` I do see logs in `/var/lib/containers/storage/overlay-containers/UUID/userdata/ctr.log` but I don't want to pull logs from there.
 
 I'm trying to add an OpenTelemetry collector to my docker compose stack and have the collector's filelog receiver watch the log files (in JSON mode is OK if required) and send them somewhere of my choosing.
